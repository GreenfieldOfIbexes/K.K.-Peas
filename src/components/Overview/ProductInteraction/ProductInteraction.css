.style-selectors {
	display: flex;
	flex-wrap: wrap;
	justify-content: space-between;
	align-items: center;
}

.style-selector {
	margin: 0 0 10px;
}

.style-image {
	height: 100px;
	width: 100px;
	border-radius: 100%;
	background-size: cover;
	background-position: center;
	background-repeat: no-repeat;
}
