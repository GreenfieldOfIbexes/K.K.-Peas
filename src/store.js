import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducers/index.js";
import { composeWithDevTools } from "redux-devtools-extension";
import promiseMiddleware from "redux-promise";

const store = createStore(
	rootReducer,
	{},
	composeWithDevTools(applyMiddleware(thunk, promiseMiddleware)),
);

export default store;
