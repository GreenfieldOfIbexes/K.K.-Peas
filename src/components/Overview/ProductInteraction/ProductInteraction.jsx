import React, { useState } from "react";
import constants from "../../../constants";
import $ from "jquery";

import "./ProductInteraction.css";

const ProductInteraction = ({
	mainProduct,
	view,
	updateStyle,
	cartHandler,
	addToOutfitHandler,
}) => {
	const currentStyle = mainProduct.styles.results[view.style_index];

	const default_size_select = "Select Size";

	// Internal state for the product options that are currently selected. The cart_clicked prop is used to conditionally render the "select a size" error message.
	const [choice, setChoice] = useState({
		size: default_size_select,
		count: "",
		cart_clicked: false,
	});

	// Utility for changing anything in internal/hooks state
	const changeState = (value, prop) => {
		setChoice((oldValues) => {
			const newValues = Object.create(oldValues);
			newValues[prop] = value;
			return newValues;
		});
	};

	// Array of the html for each style selector
	const styleSelectors = mainProduct.styles.results.map((style, index) => {
		const backgroundImageStyle = {
			backgroundImage: `url(${style.photos[0].thumbnail_url ||
				"https://www.hertrack.com/wp-content/uploads/2018/10/no-image.jpg"})`,
		};
		return (
			<div
				className="style-selector wrapper"
				key={index}
				onClick={() => {
					updateStyle(index);
					styleChangeToState(index);
				}}>
				<div className="style-image" style={backgroundImageStyle}></div>
				{view.style_index === index && (
					<div className="selected-style border">
						<div className="selected-style checkmark">
							<i className="material-icons">check</i>
						</div>
					</div>
				)}
			</div>
		);
	});

	// Array of the <option></option> tags for the size dropdown
	const sizeOptions = constants.SIZE_OPTIONS.map((size) => {
		if (currentStyle.skus[size]) {
			return (
				<option name="size" value={size} key={size}>
					{size}
				</option>
			);
		}
	});

	// Array of the <option></option> tags for the quantity dropdown
	const quantityOptions = (() => {
		var options = [];
		for (let i = 0; i < Math.min(currentStyle.skus[choice.size], 15); i++) {
			options.push(
				<option name="count" value={i + 1} key={i}>
					{i + 1}
				</option>,
			);
		}
		return options;
	})();

	const addToCartClickHandler = () => {
		changeState(true, "cart_clicked");
		if (!noProducts) {
			if (choice.size !== default_size_select) {
				cartHandler(
					{ product_id: mainProduct.id },
					window.localStorage.getItem("user_session"),
					noProducts,
				);
			} else {
				// display "please select a price"
			}
		}
	};

	// Determine whether or not there are no products
	const firstSku = Object.keys(currentStyle.skus)[0];
	const noProducts =
		firstSku === undefined || firstSku === "null" ? true : false;

	return (
		<div className="product-interaction">
			<div className="style-name">Style: {currentStyle.name}</div>
			<div className="style-selectors wrapper">{styleSelectors}</div>
			<div className="size-quantity wrapper">
				{choice.cart_clicked && choice.size === default_size_select && (
					<p className="error">Please select a size</p>
				)}
				<select
					value={choice.size}
					onChange={(e) => {
						changeState(encodeURI.target.value, "size");
						if (choice.count === "") choice.count = 1;
					}}
					className="size select">
					<option value={default_size_select}>Select Size</option>
					{sizeOptions}
					{noProducts && <p class="error">Out of stock</p>}
				</select>
				<select
					value={choice.count}
					onChange={(event) => {
						changeState(event.target.value, "count");
					}}
					className="count select"
					defaultValue="-"
					disabled={choice.size === default_size_select ? true : false}>
					<option value="-" disabled>
						-
					</option>
					{quantityOptions}
				</select>
			</div>
			<div className="cart-outfit wrapper">
				<div
					className={noProducts ? "add-to-cart no-products" : "add-to-cart"}
					onClick={addToCartClickHandler}>
					Add to Cart
				</div>
				<div
					className="add-to-outfit"
					onClick={() => {
						addToOutfitHandler(mainProduct);
						$("html, body").animate(
							{
								scrollTop: $("html, body").height(),
							},
							500,
						);
					}}>
					Add to Outfit
				</div>
			</div>
		</div>
	);
};

export default ProductInteraction;
