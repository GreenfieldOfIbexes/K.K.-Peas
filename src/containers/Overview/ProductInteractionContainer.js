import ProductInteraction from "../../components/Overview/ProductInteraction/ProductInteraction.jsx";
import newStyle from "../../actions/newStyle";
import { connect } from "react-redux";
import addToCart from "../../actions/addToCart";
import axios from "axios";
import constants from "../../constants";
import updateOutfit from "../../actions/updateOutfit.js";

/*
	mapDispatchToProps passes "dispatch" down to the functions it contains. Keeping the function definitions outside of mapDispatchToProps allows for easier unit testing and greater modularity in general.
*/

// Change the style index in store
function _updateStyle(dispatch, index) {
	dispatch(newStyle(index));
}

// Add something to the cart
function _cartHandler(dispatch, skuObj, user_session, noProducts) {
	// As long as a product is selected...
	if (noProducts === false) {
		dispatch(addToCart(skuObj));
		// Post that bad larry to the Ay Pee Ayyeeeee!!!!!
		axios.post(`${constants.API_URL}/cart`, {
			product_id: skuObj.product_id,
			user_session,
		});
		return actionObject;
	}
}

// Adding the current product to the user's saved outfit
function _addToOutfitHandler(dispatch, productObject) {
	dispatch(updateOutfit.addOutfitItem(productObject));
}

const mapStateToProps = ({ mainProduct, view }, ownProps) => ({
	mainProduct,
	view,
});

const mapDispatchToProps = (dispatch, ownProps) => ({
	updateStyle: (index) => _updateStyle(dispatch, index),
	cartHandler: (skuObj, user_session, noProducts) =>
		_cartHandler(dispatch, skuObj, user_session, noProducts),
	addToOutfitHandler: (productObject) =>
		_addToOutfitHandler(dispatch, productObject),
});

const ProductInteractionContainer = connect(
	mapStateToProps,
	mapDispatchToProps,
)(ProductInteraction);

export default ProductInteractionContainer;
