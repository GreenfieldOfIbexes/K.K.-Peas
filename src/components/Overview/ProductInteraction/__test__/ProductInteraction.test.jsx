import React from "react";
import renderer from "react-test-renderer";
import { mainProduct, view } from "../../../../../tests/initialStateStub";
import ProductInteraction from "../ProductInteraction";
import ProductInteractionContainer from "../../../../containers/Overview/ProductInteractionContainer";
import store from "../../../../store";

it("Renders properly on load", () => {
	const renderedComponent = renderer
		.create(
			<ProductInteractionContainer mainProduct={mainProduct} view={view} />,
		)
		.toJSON();
	expect(renderedComponent).toMatchSnapshot();
});

it("Selecting a new style only updates style_index", () => {
	const viewStyleIndex1 = Object.assign({}, view, { style_index: 1 });

	const renderedComponent = renderer
		.create(
			<ProductInteractionContainer
				mainProduct={mainProduct}
				view={viewStyleIndex1}
			/>,
		)
		.toJSON();
	expect(renderedComponent).toMatchSnapshot();
});

// Props ProductInteraction needs:
// mainProduct,
// view,
// updateStyle,
// cartHandler,
// addToOutfitHandler,
