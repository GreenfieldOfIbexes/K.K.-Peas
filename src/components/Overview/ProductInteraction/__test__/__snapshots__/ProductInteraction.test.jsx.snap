// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Interaction renders correctly 1`] = `
<div
  className="product-interaction"
>
  <div
    className="style-name"
  >
    Style: 
    Forest Green & Black
  </div>
  <div
    className="style-selectors wrapper"
  >
    <div
      className="style-selector wrapper"
      onClick={[Function]}
    >
      <div
        className="style-image"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1501088430049-71c79fa3283e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80)",
          }
        }
      />
      <div
        className="selected-style border"
      >
        <div
          className="selected-style checkmark"
        >
          <i
            className="material-icons"
          >
            check
          </i>
        </div>
      </div>
    </div>
    <div
      className="style-selector wrapper"
      onClick={[Function]}
    >
      <div
        className="style-image"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1560567546-4c6dbc16877b?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80)",
          }
        }
      />
    </div>
    <div
      className="style-selector wrapper"
      onClick={[Function]}
    >
      <div
        className="style-image"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1556304653-cba65c59b3c5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80)",
          }
        }
      />
    </div>
    <div
      className="style-selector wrapper"
      onClick={[Function]}
    >
      <div
        className="style-image"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1530092376999-2431865aa8df?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80)",
          }
        }
      />
    </div>
    <div
      className="style-selector wrapper"
      onClick={[Function]}
    >
      <div
        className="style-image"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1448526478325-616f2b15b04e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80)",
          }
        }
      />
    </div>
    <div
      className="style-selector wrapper"
      onClick={[Function]}
    >
      <div
        className="style-image"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1514866726862-0f081731e63f?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80)",
          }
        }
      />
    </div>
  </div>
  <div
    className="size-quantity wrapper"
  >
    <select
      className="size select"
      displayEmpty={true}
      onChange={[Function]}
      value="Select Size"
    >
      <option
        value="Select Size"
      >
        Select Size
      </option>
      <option
        name="size"
        value="XS"
      >
        XS
      </option>
      <option
        name="size"
        value="S"
      >
        S
      </option>
      <option
        name="size"
        value="M"
      >
        M
      </option>
      <option
        name="size"
        value="L"
      >
        L
      </option>
      <option
        name="size"
        value="XL"
      >
        XL
      </option>
    </select>
    <select
      className="count select"
      disabled={true}
      displayEmpty={true}
      onChange={[Function]}
      value=""
    >
      <option
        disabled={true}
        selected={true}
        value="-"
      >
        -
      </option>
    </select>
  </div>
  <div
    className="cart-outfit wrapper"
  >
    <div
      className="add-to-cart"
      onClick={[Function]}
    >
      Add to Cart
    </div>
    <div
      className="add-to-outfit"
      onClick={[Function]}
    >
      Add to Outfit
    </div>
  </div>
</div>
`;
